### ISSUE TITLE
Using Grad-CAM with YOLOv8 Classification Model — Is This Implementation Correct?

### ISSUE THREAD

**Issue by Cestovatels** [ultralytics/ultralytics#23819]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello. I tried to create an example of using Gradcam for the Yolov8x classification model. The results I'm getting so far seem quite accurate. Does this make sense? Thank you in advance for your answers. @Y-T-G  

````py
import cv2
import numpy as np
import torch
import torch.nn as nn
from ultralytics import YOLO
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image

class YOLOClsWrapper(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model

    def forward(self, x):
        res = self.model(x)
        return res[0] if isinstance(res, (list, tuple)) else res


model_path = r'best.pt' #(classification model)
image_path = r'test.jpg'

base_model = YOLO(model_path)
wrapped_model = YOLOClsWrapper(base_model.model)
wrapped_model.eval()


img = cv2.imread(image_path)
# img = cv2.resize(img, (224, 224))
rgb_img = np.float32(img) / 255
input_tensor = torch.from_numpy(np.transpose(rgb_img, (2, 0, 1))).unsqueeze(0).to(torch.float32)


input_tensor.requires_grad = True 


target_layers = [base_model.model.model[-2]]


cam = GradCAM(model=wrapped_model, target_layers=target_layers)


grayscale_cam = cam(input_tensor=input_tensor, targets=None)
grayscale_cam = grayscale_cam[0, :]


cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)
cv2.imwrite('grad_cam_final.jpg', cv2.cvtColor(cam_image, cv2.COLOR_RGB2BGR))
````

<img width="202" height="152" alt="Image" src="https://github.com/user-attachments/assets/d1b4fe6a-a131-4438-88ae-d44109abdb55" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Cestovatels, thanks for opening this issue and sharing your Grad-CAM classification example 🚀 This is an automated response to help with triage, and an Ultralytics engineer will also assist you soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Yes, the overall approach is reasonable (wrapping the raw `base_model.model` so Grad-CAM can backprop), and targeting the penultimate layer is a sensible default, but two details in your snippet are likely to skew/soften the heatmaps: you’re feeding unnormalized inputs (Ultralytics classification uses its own transforms in `ClassificationPredictor.preprocess`) and your wrapper is returning probabilities (post-`softmax`) instead of logits (better Grad-CAM signal). See the [classification predictor preprocessing reference](https://docs.ultralytics.com/reference/models/yolo/classify/predict/) and the [classification head behavior (returns `(probs, logits)` in eval)](https://docs.ultralytics.com/reference/nn/modules/head/) for context, plus the [Ultralytics Grad-CAM/XAI overview](https://www.ultralytics.com/blog/all-you-need-to-know-about-explainable-ai).

```python
class YOLOClsWrapper(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model

    def forward(self, x):
        out = self.model(x)
        if isinstance(out, (list, tuple)) and len(out) == 2:  # (probs, logits)
            return out[1]  # logits for Grad-CAM
        return out

# make sure your input matches the model's expected classify transforms (resize + normalization),
# otherwise CAMs can look "right" but be inconsistent across images/models.
```

If you share what `base_model.model.model[-2]` actually is for your `best.pt` (print the model) and your training `imgsz`, I can confirm the most appropriate `target_layers` for that exact checkpoint.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@Cestovatels, #25548 from @Y-T-G is now merged, so classification CAMs no longer require a custom model wrapper or a manually selected target layer. `model.predict(..., visualize=True)` uses the classification head path and saves the resulting LayerCAM overlay directly.
